### PR TITLE
Refactor core SDK code into Microsoft.Graph.Core project

### PR DESCRIPTION
**Pull request by ginach**
The vast majority of these changes are purely a refactor into different projects (and a re-gen based on latest template changes). I tried to make these changes in a way where developers wouldn't notice. The one exception is that some constants from Constants.cs now live in CoreConstants.cs. Keeping the names the same caused more issues than it fixed and the affected constants most likely aren't in use outside the SDK code.

The other big change is to DerivedTypeConverter.cs. Since the type we're resolving to doesn't live in the same binary anymore (and we're limited to the reflection abilities of .NET PCL) we now use the following logic to resolve the type:
    - Is the requested type in the assembly of the type the deserializer wants to use? If so, create that type.
    - If not, create an instance of the type the deserializer wants to use.

Ideally, DerivedTypeConverter would be able to take in an Assembly or something similar to specify where the type lives but that requires a higher version of JSON.NET (and can be done later as non-breaking).

We also now cache the type resolution mappings. This way we're not doing the various type -> assembly lookup every time the API returns "microsoft.graph.user", for example.

### CODE REVIEW

File: src/Microsoft.Graph.Core/Requests/BaseRequest.cs [hunk 1]
```diff
@@ -0,0 +1,343 @@
+// ------------------------------------------------------------------------------
```
Comment (shiftylogic) [inline src/Microsoft.Graph.Core/Requests/BaseRequest.cs]: Is this just moving from another place? If so, why did Git treat this as a new file. It should be able to handle move operations.

File: src/Microsoft.Graph.Core/Requests/BaseRequest.cs [hunk 2]
```diff
@@ -40,6 +41,9 @@ public class BaseRequest : IBaseRequest
 
             this.RequestUrl = this.InitializeUrl(requestUrl);
 
+            this.sdkVersionHeaderName = CoreConstants.Headers.SdkVersionHeaderName;
```
Comment (shiftylogic) [inline src/Microsoft.Graph.Core/Requests/BaseRequest.cs]: Should these be changeable at run-time. If not, I'd make them const in this class.

File: src/Microsoft.Graph.Core/Requests/BaseRequest.cs [hunk 3]
```diff
@@ -296,7 +309,7 @@ private string InitializeUrl(string requestUrl)
                 throw new ServiceException(
                     new Error
                     {
-                        Code = GraphErrorCode.InvalidRequest.ToString(),
+                        Code = "invalidRequest",
```
Comment (shiftylogic) [inline src/Microsoft.Graph.Core/Requests/BaseRequest.cs]: Why the move to hardcoded strings? I'd advocate that these should be in a constants file somewhere.

File: src/Microsoft.Graph.Core/Requests/BaseRequest.cs (hunk 2 above)
Comment (ginach) [inline src/Microsoft.Graph.Core/Requests/BaseRequest.cs]: Yes, they should. The specific use case right now is we're going to consume this core project in the OneDrive SDK to reduce duplicate code and those telemetry headers are completely different. That said, we want to encourage the newer format for other SDKs. That's why these are standard variables while prefix is a property.

File: src/Microsoft.Graph.Core/Serialization/DerivedTypeConverter.cs [hunk 4]
```diff
@@ -119,8 +152,8 @@ private object Create(string typeString)
                 throw new ServiceException(
                     new Error
                     {
-                        Code = GraphErrorCode.GeneralException.ToString(),
-                        Message = string.Format("Unable to create an instance of type {0}.", typeString),
+                        Code = "generalException",
+                        Message = string.Format("Unable to create an instance of type {0}.", type.FullName),
```
Comment (shiftylogic) [inline src/Microsoft.Graph.Core/Serialization/DerivedTypeConverter.cs]: The strings here should be in a string resource file for possible localization.

**Pull request merged**